### PR TITLE
Implement workflow execution recovery

### PR DESCRIPTION
**Pull request by katrogan**
# TL;DR
Implement workflow execution recovery. For ordinary task nodes this modifies pre-execute to attempt to recover node executions when the overall workflow execution is run in recovery mode, helping elide needless re-computation for previously succeeded node executions.

For workflow nodes (that is, those that call out to create an execution from a launch plan) this sets the triggered executions to recover from the original node execution which created an execution using the same launch plan.

For dynamic nodes (that is those that contain a subworkflow) the behavior remains the same as task nodes. If the original dynamic node succeeded, great, we'll recover it. Otherwise the dynamic workflow runs again.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1151

## Follow-up issue
_NA_

### CODE REVIEW

File: pkg/apis/flyteworkflow/v1alpha1/iface.go [hunk 1]
```diff
@@ -92,6 +94,10 @@ func (p NodePhase) String() string {
 		return "RetryableFailure"
 	case NodePhaseDynamicRunning:
 		return "DynamicRunning"
+	case NodePhaseRecovering:
```
Comment (kumare3) [inline pkg/apis/flyteworkflow/v1alpha1/iface.go:97]: you have to do op_code_generate

File: pkg/controller/nodes/subworkflow/launchplan/admin_test.go [hunk 2]
```diff
@@ -193,6 +195,42 @@ func TestAdminLaunchPlanExecutor_Launch(t *testing.T) {
 		assert.NoError(t, err)
 	})
 
+	t.Run("happy recover", func(t *testing.T) {
```
Comment (kumare3) [inline pkg/controller/nodes/subworkflow/launchplan/admin_test.go:198]: what if recover fails?

File: pkg/controller/nodes/executor.go [hunk 3]
```diff
@@ -146,6 +152,119 @@ func (c *nodeExecutor) IdempotentRecordEvent(ctx context.Context, nodeEvent *eve
...
+	if recoveredData.FullOutputs != nil {
+		outputs = recoveredData.FullOutputs
+	} else if len(recovered.Closure.GetOutputUri()) > 0 {
+		if err := c.store.ReadProtobuf(ctx, storage.DataReference(recovered.Closure.GetOutputUri()), outputs); err != nil {
+			return handler.PhaseInfoUndefined, errors.Wrapf(errors.InputsNotFoundError, nCtx.NodeID(), err, "failed to read output data [%v].", recovered.Closure.GetOutputUri())
+		}
+	} else {
+		logger.Debugf(ctx, "No outputs found for recovered node [%+v]", nCtx.NodeExecutionMetadata().GetNodeExecutionID())
+	}
+	outputFile := v1alpha1.GetOutputsFile(nCtx.NodeStatus().GetOutputDir())
```
Comment (kumare3) [inline pkg/controller/nodes/executor.go:234]: ohh no are we still using this?

File: pkg/controller/nodes/executor.go [hunk 4]
```diff
@@ -429,14 +554,19 @@ func (c *nodeExecutor) handleQueuedOrRunningNode(ctx context.Context, nCtx *node
 		np = v1alpha1.NodePhaseSucceeded
 		finalStatus = executors.NodeStatusSuccess
 	}
+	if np == v1alpha1.NodePhaseRecovering && !h.FinalizeRequired() {
```
Comment (kumare3) [inline pkg/controller/nodes/executor.go:557]: if it is recovering we do not even care about the  handler or finalization right?

Comment (kumare3) [reply pkg/controller/nodes/executor.go:557]: do we really need `Recovering`?

**Comment by kumare3** [comment]
Its amazing how clean this change is. It really fits in well into the premise?

File: pkg/controller/controller.go [hunk 5]
```diff
@@ -419,9 +422,20 @@ func New(ctx context.Context, cfg *config.Config, kubeclientset kubernetes.Inter
 
 	controller.levelMonitor = NewResourceLevelMonitor(scope.NewSubScope("collector"), flyteworkflowInformer.Lister())
 
+	// The admin client might not be initialized if EnableAdminLauncher is set to False.
+	if adminClient == nil {
+		var err error
+		adminClient, err = getAdminClient(ctx)
```
Comment (EngHabu) [inline pkg/controller/controller.go:428]: If we now always want an admin client, let's remove the condition and just always initialize it and use it up there if needed and down here all the time...

File: pkg/controller/nodes/subworkflow/launchplan.go [hunk 6]
```diff
@@ -17,7 +21,8 @@ import (
 )
 
 type launchPlanHandler struct {
-	launchPlan launchplan.Executor
+	launchPlan     launchplan.Executor
+	recoveryClient recovery.Client
```
Comment (EngHabu) [inline pkg/controller/nodes/subworkflow/launchplan.go:25]: I'm a bit confused about why do we need the recovery client at this layer... won't that be handled at the Node executor layer?

Comment (kumare3) [reply pkg/controller/nodes/subworkflow/launchplan.go:25]: Isn't the reason that we want to call the recover execution endpoint instead in case the child workflow node has failed, so that we can recover a partially failed child workflow?

Comment (EngHabu) [reply pkg/controller/nodes/subworkflow/launchplan.go:25]: That's what I thought.. but looking at the interface and implementation of recovery client... it retrieves Node executions from admin... I think... I obviously could be missing something since I'm half looking at this and half prepping my bags :-D

File: pkg/controller/nodes/executor.go (hunk 3 above)
Comment (katrogan) [reply pkg/controller/nodes/executor.go:234]: @kumare3 what should i be using instead?

File: pkg/controller/nodes/subworkflow/launchplan.go (hunk 6 above)
Comment (katrogan) [reply pkg/controller/nodes/subworkflow/launchplan.go:25]: yes, we retrieve the node execution - which has target metadata of type workflow node, which we can use to fetch the originally-launched child execution

File: pkg/controller/nodes/executor.go (hunk 4 above)
Comment (katrogan) [reply pkg/controller/nodes/executor.go]: updated

File: pkg/controller/nodes/subworkflow/launchplan/admin_test.go (hunk 2 above)
Comment (katrogan) [reply pkg/controller/nodes/subworkflow/launchplan/admin_test.go:198]: updated to call create execution when applicable

File: pkg/controller/controller.go (hunk 5 above)
Comment (katrogan) [reply pkg/controller/controller.go]: done

**Comment by katrogan** [comment]
PTAL @kumare3

File: pkg/controller/nodes/executor.go (hunk 3 above)
Comment (kumare3) [reply pkg/controller/nodes/executor.go:234]: i think we use this now - https://github.com/flyteorg/flyteplugins/blob/60b94c688ef2b98aa53a9224b529ac672af04540/go/tasks/pluginmachinery/ioutils/paths.go

Comment (katrogan) [reply pkg/controller/nodes/executor.go:234]: @kumare3 this doesn't include output file paths though. should i update it?

**Review by kumare3 — APPROVED** [review]
